File: core/src/main/java/pcb/uwu/amount/derived/finance/InterestRate.java
package pcb.uwu.amount.derived.finance;

import org.jetbrains.annotations.NotNull;
import pcb.uwu.amount.finance.Money;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.unit.derived.fundamental.FrequencyUnit;
import pcb.uwu.unit.finance.CurrencyUnit;
import pcb.uwu.unit.finance.RentUnit;

import java.math.BigDecimal;
import java.math.MathContext;

import static pcb.uwu.utils.UnitAmountUtils.dividedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.getAmountIn;
import static pcb.uwu.utils.UnitAmountUtils.minusAmount;
import static pcb.uwu.utils.UnitAmountUtils.multipliedByScalar;
import static pcb.uwu.utils.UnitAmountUtils.plusAmount;

public class InterestRate extends CompositeUnitAmount<FrequencyUnit> {

	// region constructors

	public InterestRate(Number value, FrequencyUnit unit) {
		super(value, unit);
	}

	public InterestRate(Number value, Magnitude magnitude, FrequencyUnit unit) {
		super(value, magnitude, unit);
	}

	public InterestRate(String value, FrequencyUnit unit) {
		super(value, unit);
	}

	public InterestRate(String value, Magnitude magnitude, FrequencyUnit unit) {
		super(value, magnitude, unit);
	}

	public InterestRate(BigDecimal value, FrequencyUnit unit) {
		super(value, unit);
	}

	public InterestRate(BigDecimal value, Magnitude magnitude, FrequencyUnit unit) {
		super(value, magnitude, unit);
	}

	public InterestRate(BigDecimalAmount amount, FrequencyUnit unit) {
		super(amount, unit);
	}

	public InterestRate(BigDecimalAmount amount, Magnitude magnitude, FrequencyUnit unit) {
		super(amount, magnitude, unit);
	}

	// endregion

	// region implement UnitAmount

	@NotNull
	@Override
	public InterestRate plus(@NotNull UnitAmount<FrequencyUnit> other) {
		return new InterestRate(plusAmount(this, other), getUnit());
	}

	@NotNull
	@Override
	public InterestRate minus(@NotNull UnitAmount<FrequencyUnit> other) {
		return new InterestRate(minusAmount(this, other), getUnit());
	}

	@Override
	public InterestRate multiply(BigDecimal other, MathContext mathContext) {
		return new InterestRate(multipliedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public InterestRate div(BigDecimal other, MathContext mathContext) {
		return new InterestRate(dividedByScalar(this, other, mathContext), getUnit());
	}

	@Override
	public InterestRate into(FrequencyUnit unit) {
		return new InterestRate(getAmountIn(this, unit), unit);
	}

	// endregion

	// region composition

	public Rent multipliedBy(Money money, MathContext mathContext) {
		BigDecimalAmount amount = multipliedByScalar(this, money.getAmount().getValue(), mathContext);
		RentUnit unit = new RentUnit(money.getUnit(), getUnit());

		return new Rent(amount, unit);
	}

	public Money multipliedBy(Debt debt, MathContext mathContext) {
		BigDecimalAmount amount = super.multiply(debt, mathContext).getAmount();
		CurrencyUnit unit = getUnit().getUnitCounter().findUnit(CurrencyUnit.class);

		return new Money(amount, unit);
	}

	// endregion
}

File: core/src/main/java/pcb/uwu/amount/derived/finance/InterestRate.kt
package pcb.uwu.amount.derived.finance

import pcb.uwu.amount.base.Currency
import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.CurrencyUnit
import pcb.uwu.unit.derived.fundamental.FrequencyUnit
import pcb.uwu.unit.finance.RentUnit
import pcb.uwu.utils.UnitAmountUtils

class InterestRate : CompositeUnitAmount<FrequencyUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: FrequencyUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: FrequencyUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(interestRate: UnitAmount<FrequencyUnit>) =
        InterestRate(amount = this.amount + (interestRate to this.unit).amount,
                     unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(interestRate: UnitAmount<FrequencyUnit>) =
        InterestRate(amount = this.amount - (interestRate to this.unit).amount,
                     unit = this.unit)

    override fun times(scalar: Number) =
        InterestRate(amount = this.amount * scalar,
                     unit = this.unit)

    override fun div(scalar: Number) =
        InterestRate(amount = this.amount / scalar,
                     unit = this.unit)

    override fun to(unit: FrequencyUnit) =
        InterestRate(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
                     unit = unit)

    // endregion

    // region composition

    fun times(currency: Currency) =
        Rent(amount = UnitAmountUtils.times(this, currency.amount.value),
             unit = RentUnit(interestRateUnit = this.unit,
                             currency = currency.unit))

    fun times(debt: Debt) =
        Currency(amount = (this * debt).amount,
                 unit = this.unit.unitCounter.findUnit(CurrencyUnit::class.java)!!)

    // endregion
}